File: server/internal/api/audit.go
package api

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/serversupervisor/server/internal/config"
	"github.com/serversupervisor/server/internal/database"
	"github.com/serversupervisor/server/internal/models"
)


type AuditHandler struct {
	db  *database.DB
	cfg *config.Config
}

func NewAuditHandler(db *database.DB, cfg *config.Config) *AuditHandler {
	return &AuditHandler{db: db, cfg: cfg}
}

// GetAuditLogs returns paginated audit logs (admin only)
func (h *AuditHandler) GetAuditLogs(c *gin.Context) {
	// Check if user is admin
	if c.GetString("role") != models.RoleAdmin {
		c.JSON(http.StatusForbidden, gin.H{"error": "insufficient permissions"})
		return
	}

	// Get pagination params
	page := 1
	limit := 50
	if p := c.Query("page"); p != "" {
		if parsed, err := strconv.Atoi(p); err == nil && parsed > 0 {
			page = parsed
		}
	}
	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil && parsed > 0 && parsed <= 100 {
			limit = parsed
		}
	}

	offset := (page - 1) * limit

	logs, err := h.db.GetAuditLogs(limit, offset)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch audit logs"})
		return
	}

	if logs == nil {
		logs = []models.AuditLog{}
	}

	c.JSON(http.StatusOK, gin.H{
		"logs":  logs,
		"page":  page,
		"limit": limit,
	})
}

// GetAuditLogsByHost returns audit logs for a specific host
func (h *AuditHandler) GetAuditLogsByHost(c *gin.Context) {
	hostID := c.Param("host_id")
	if hostID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "host_id required"})
		return
	}

	// Get limit param
	limit := 100
	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil && parsed > 0 && parsed <= 500 {
			limit = parsed
		}
	}

	logs, err := h.db.GetAuditLogsByHost(hostID, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch audit logs"})
		return
	}

	if logs == nil {
		logs = []models.AuditLog{}
	}

	c.JSON(http.StatusOK, logs)
}

// GetMyAuditLogs returns the current user's own audit logs
func (h *AuditHandler) GetMyAuditLogs(c *gin.Context) {
	username := c.GetString("username")
	if username == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	limit := 10
	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil && parsed > 0 && parsed <= 100 {
			limit = parsed
		}
	}

	logs, err := h.db.GetAuditLogsByUser(username, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch logs"})
		return
	}
	if logs == nil {
		logs = []models.AuditLog{}
	}

	c.JSON(http.StatusOK, gin.H{
		"user": username,
		"logs": logs,
	})
}

// GetCommandsHistory returns paginated remote commands for all hosts (admin and operator).
func (h *AuditHandler) GetCommandsHistory(c *gin.Context) {
	role := c.GetString("role")
	if role != models.RoleAdmin && role != models.RoleOperator {
		c.JSON(http.StatusForbidden, gin.H{"error": "insufficient permissions"})
		return
	}

	page, limit := 1, 50
	if p := c.Query("page"); p != "" {
		if parsed, err := strconv.Atoi(p); err == nil && parsed > 0 {
			page = parsed
		}
	}
	offset := (page - 1) * limit

	cmds, err := h.db.GetAllRemoteCommands(limit, offset)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch commands history"})
		return
	}
	total, _ := h.db.CountAllRemoteCommands()
	if cmds == nil {
		cmds = []database.RemoteCommandWithHost{}
	}
	c.JSON(http.StatusOK, gin.H{"commands": cmds, "total": total, "page": page, "limit": limit})
}

// GetAuditLogsByUser returns audit logs for a specific user (admin only)
func (h *AuditHandler) GetAuditLogsByUser(c *gin.Context) {
	// Check if user is admin
	if c.GetString("role") != models.RoleAdmin {
		c.JSON(http.StatusForbidden, gin.H{"error": "insufficient permissions"})
		return
	}

	username := c.Param("username")
	if username == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "username required"})
		return
	}

	limit := 100
	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil && parsed > 0 && parsed <= 500 {
			limit = parsed
		}
	}

	logs, err := h.db.GetAuditLogsByUser(username, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch logs"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"user": username,
		"logs": logs,
	})
}

File: frontend/src/api/audit.ts
import { api } from './client'
import type { AuditLog, RemoteCommand, RemoteCommandWithHost, HostTimelineEvent } from '../types/audit'

export interface AuditLogFilters {
  category?: string
  from?: string
  to?: string
}

export const auditApi = {
  getAuditLogs: (page?: number, limit?: number, filters?: AuditLogFilters) =>
    api.get<{ logs: AuditLog[], page: number, limit: number }>('/v1/audit/logs', { params: { page: page ?? 1, limit: limit ?? 50, ...filters } }),
  exportAuditLogs: (filters?: AuditLogFilters) =>
    api.get('/v1/audit/logs/export', { params: filters, responseType: 'blob' }),
  getMyAuditLogs: (limit?: number) => api.get<AuditLog[]>('/v1/audit/logs/me', { params: { limit: limit ?? 10 } }),
  getAuditLogsByHost: (hostId: string, limit?: number) =>
    api.get<AuditLog[]>(`/v1/audit/logs/host/${hostId}`, { params: { limit: limit ?? 100 } }),
  getAuditLogsByUser: (username: string, limit?: number) =>
    api.get<{ user: string, logs: AuditLog[] }>(`/v1/audit/logs/user/${username}`, { params: { limit: limit ?? 100 } }),
  getCommandsHistory: (page?: number, limit?: number, filters?: { search?: string; module?: string; status?: string }, signal?: AbortSignal) =>
    api.get<{ commands: RemoteCommandWithHost[], total: number, page: number, limit: number }>('/v1/audit/commands', { params: { page: page ?? 1, limit: limit ?? 50, ...filters }, signal }),
  getCommandStatus: (id: string) => api.get<RemoteCommand>(`/v1/commands/${id}`),
  cancelCommand: (id: string) => api.post<{ status: string }>(`/v1/commands/${id}/cancel`),
  getHostTimeline: (hostId: string, limit?: number) =>
    api.get<{ events: HostTimelineEvent[] }>(`/v1/hosts/${hostId}/timeline`, { params: { limit: limit ?? 50 } }),
}
